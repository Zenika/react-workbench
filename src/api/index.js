const bodyParser = require('body-parser')
const { API_BASE_CONTEXT } = require('../constants')
const state = require('./models/state')

const genPath = resource => `${API_BASE_CONTEXT}/${resource.NAME}`

const connect = (app) => {
  // use json for api
  app.use(API_BASE_CONTEXT, bodyParser.json())

  // connects services
  // 1. State
  app.post(genPath(state), async (req, res) => {
    try {
      await state.create(req.body)
      res.sendStatus(200)
    } catch (ex) {
      // TODO
      // eslint-disable-next-line no-console
      console.error(ex)
      res.sendStatus(500)
    }
  })
}

module.exports = {
  connect,
}
