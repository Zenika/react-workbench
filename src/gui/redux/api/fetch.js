import { readme, html } from '../store'

const fetchDoc = reducer => url => async (dispatch) => {
  dispatch(reducer.reset())
  try {
    const response = await fetch(url)
    const text = await response.text()
    dispatch(reducer.set(text))
  } catch (error) {
    dispatch(reducer.set(''))
  }
}

export default {
  readme: () => fetchDoc(readme)('/api/readme'),
  html: () => fetchDoc(html)('/api/html'),
}
