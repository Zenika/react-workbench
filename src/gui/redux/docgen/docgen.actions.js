export const REQUEST_DOCGEN = 'REQUEST_DOCGEN'
export const RECEIVED_DOCGEN = 'RECEIVED_DOCGEN'
export const RECEIVED_EMPTY_DOCGEN = 'RECEIVED_EMPTY_DOCGEN'
export const FAILED_DOCGEN = 'FAILED_DOCGEN'

export const fetchDocgen = () => async (dispatch) => {
  dispatch({ type: REQUEST_DOCGEN })
  try {
    const response = await fetch('/api/docgen')
    const json = await response.json()
    if (json && json.length > 0) {
      dispatch({ type: RECEIVED_DOCGEN, payload: json[0] })
    } else {
      dispatch({ type: RECEIVED_EMPTY_DOCGEN })
    }
  } catch (error) {
    dispatch({ type: FAILED_DOCGEN, error })
  }
}
