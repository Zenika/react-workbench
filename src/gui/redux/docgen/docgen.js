import factory from 'trampss-redux-factory'

export default factory({ name: 'docgen', type: 'uniq' })
