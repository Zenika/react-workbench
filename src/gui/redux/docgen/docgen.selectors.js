import { createSelector } from 'reselect'
import { docgen } from '../reducers'

const getProps = state => docgen.get()(state).props || {}
export const getPropsKeys = createSelector(getProps, p => Object.keys(p))
export const getProp = name => createSelector(getProps, p => p[name])
