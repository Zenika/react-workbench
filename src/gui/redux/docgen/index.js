export { default } from './docgen'
export * from './docgen.selectors'
export * from './docgen.api'
