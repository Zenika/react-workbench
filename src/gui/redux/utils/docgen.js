const convertDocgenValue = (value, type) => {
  switch (type) {
    case 'bool':
      return value === 'true'
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'symbol':
      return value
    default:
      return eval(value) // eslint-disable-line no-eval
  }
}

const docgenToModel = (docgen) => {
  return Object.keys(docgen.props || {}).reduce((obj, key) => {
    const prop = docgen.props[key]
    const type = prop.type && prop.type.name
    const value = prop.defaultValue && prop.defaultValue.value
    return { ...obj, [key]: convertDocgenValue(value, type) }
  }, {})
}

const convertToGuiValue = (value, type) => {
  switch (type) {
    case 'array':
    case 'object':
      return JSON.stringify(value)
    case 'func':
      return eval(value) // eslint-disable-line no-eval
    default:
      return value
  }
}

const convertFromGuiValue = (value, type) => {
  switch (type) {
    case 'bool':
      return !!value
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'array':
    case 'func':
      return eval(value) // eslint-disable-line no-eval
    default:
      return value
  }
}

export { convertDocgenValue, docgenToModel, convertToGuiValue, convertFromGuiValue }
