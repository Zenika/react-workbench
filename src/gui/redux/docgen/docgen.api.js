import docgen from './docgen'

export const fetchDocgen = () => async (dispatch) => {
  dispatch(docgen.reset())
  try {
    const response = await fetch('/api/doc')
    const json = await response.json()
    if (json && json.length > 0) {
      dispatch(docgen.set(json[0]))
    } else {
      dispatch(docgen.set({}))
    }
  } catch (error) {
    dispatch(docgen.set({}))
  }
}
