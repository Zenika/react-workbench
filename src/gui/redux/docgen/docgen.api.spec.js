/* eslint-env jest */
import { fetchDocgen } from './docgen.api'

const dispatch = jest.fn()

describe('gui/redux/docgen', () => {
  beforeEach(() => {
    dispatch.mockClear()
  })

  it('should dispatch RESET_DOCGEN & SET_DOCGEN with response', async () => {
    // mocks
    window.fetch = jest.fn(async () => ({
      json: async () => [{ foo: 'bar' }],
    }))

    // execute
    await fetchDocgen()(dispatch)

    // asserts
    expect(dispatch.mock.calls).toMatchSnapshot()
  })

  it('should dispatch SET_DOCGEN with empty object when no data fetched', async () => {
    // mocks
    window.fetch = jest.fn(async () => ({
      json: async () => undefined,
    }))

    // execute
    await fetchDocgen()(dispatch)

    // asserts
    expect(dispatch.mock.calls).toMatchSnapshot()
  })
  it('should dispatch SET_DOCGEN with empty object when an error occured', async () => {
     // mocks
    window.fetch = jest.fn(async () => ({
      json: async () => {
        throw Error()
      },
    }))

    // execute
    await fetchDocgen()(dispatch)

    // asserts
    expect(dispatch.mock.calls).toMatchSnapshot()
  })
})
