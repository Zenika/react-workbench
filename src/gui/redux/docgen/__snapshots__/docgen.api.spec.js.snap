// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gui/redux/docgen should dispatch RESET_DOCGEN & SET_DOCGEN with response 1`] = `
Array [
  Array [
    Object {
      "type": "@trampss/RESET_DOCGEN",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "foo": "bar",
      },
      "type": "@trampss/SET_DOCGEN",
    },
  ],
]
`;

exports[`gui/redux/docgen should dispatch SET_DOCGEN with empty object when an error occured 1`] = `
Array [
  Array [
    Object {
      "type": "@trampss/RESET_DOCGEN",
    },
  ],
  Array [
    Object {
      "payload": Object {},
      "type": "@trampss/SET_DOCGEN",
    },
  ],
]
`;

exports[`gui/redux/docgen should dispatch SET_DOCGEN with empty object when no data fetched 1`] = `
Array [
  Array [
    Object {
      "type": "@trampss/RESET_DOCGEN",
    },
  ],
  Array [
    Object {
      "payload": Object {},
      "type": "@trampss/SET_DOCGEN",
    },
  ],
]
`;
